h/security/policy/_api.py

```diff
@@ -0,0 +1,35 @@
+from pyramid.request import Request, RequestLocalCache
+
+from h.security.identity import Identity
+from h.security.policy._identity_base import IdentityBasedPolicy
+
+
+class APIPolicy(IdentityBasedPolicy):
+    """The security policy for API requests. Delegates to subpolicies."""
+
+    def __init__(self, sub_policies):
+        self.sub_policies = sub_policies
+        self._identity_cache = RequestLocalCache(self._load_identity)
+
+    def forget(self, *_args, **_kwargs):
+        raise RuntimeError("forget() not supported for stateless API requests")
+
+    def identity(self, request) -> Identity | None:
+        return self._identity_cache.get_or_create(request)
+
+    def remember(self, *_args, **_kwargs):
+        raise RuntimeError("remember() not supported for stateless API requests")
+
+    def _load_identity(self, request: Request) -> Identity | None:
+        for policy in applicable_policies(request, self.sub_policies):
+            identity = policy.identity(request)
+            if identity:
+                return identity
+
+        return None
+
+
+def applicable_policies(request, policies):
+    """Return the security policies from `policies` that can handle `request`."""
+
+    return [policy for policy in policies if policy.handles(request)]
```

Comment:
This new `APIPolicy` is used for API requests only. It delegates `identity()` method calls to a list of sub-policies in turn and goes with the first sub-policy whose `handles(request)` method returns `True` and whose `identity(request)` method returns a truthy value.

Notes:

* This inherits from [`IdentityBasedPolicy`](https://github.com/hypothesis/h/blob/d8e75ce9ca51331f721029f99cdc2cde5fa396c1/h/security/policy/_identity_base.py#L7-L47) so it inherits `authenticated_userid()` and `permits()` methods that compute their results based on what the `identity()` method returns.
* This includes hard-coded `remember()` and `forget()` methods that raise exceptions: `remember()` and `forget()` aren't supported for stateless API endpoints. There's a risk here because on `main` API policies instead have `remember()` and `forget()` methods that just `return []`, so it's possible that we're sometimes calling `remember()` or `forget()` during API requests which will now crash.
* `RequestLocalCache` is used so that `identity()` is called only once per request and the return value is cached and re-used within that request. This saves some CPU cycles, and also makes sure that multiple calls to `identity()` don't potentially delegate to different sub-policies (I don't think this could have happened anyway, but caching is a simple way to be sure of it).